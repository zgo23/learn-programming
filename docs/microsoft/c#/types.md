# in vs out vs ref

-   As you know, value types are normally passed by value, meaning a copy has to be made when passing a value as an argument. The **in** keyword enables us to avoid this copy by passing a reference instead.
-   You should only use **in** for types that are larger than a pointer.
-   You should use **in** only with readonly value types, because mutable value types can undo the performance benefits. (Mutable value types are typically a bad idea in any case.)
-   You can use the **out** and **ref** keywords with reference types too. That may sound redundant, but it can be useful. It provides double indirection—the method receives a reference to a variable that holds a reference.
-   you cannot put a reference to a local variable in something that might outlive the variable it refers to. So **you cannot use this (ref) keyword on a field**.
-   C# nested classes are equivalent to Java **static** nested classes
-   In fact, an enum is really just a fancy way of defining a load of const fields.
-   The compiler generates a fairly ordinary class definition for each anonymous type.
-   you’re free to use type parameters as arguments for other generic types.
-   Generic type ---Type arguments are supplied---> Constructed type
-   if you want to use one of your own type parameters as the type argument for a generic that specifies a constraint, you’ll need to specify the same constraint on your own type parameter.
-   All delegate types derive from System.Delegate, and all enumeration types derive from System.Enum.
-   a new array of any type will have all of its elements initialized to a known value.
