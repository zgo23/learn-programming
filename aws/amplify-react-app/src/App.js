import React, { useState, useEffect, useCallback } from "react";
import { API } from "aws-amplify";
import "./App.css";

function App() {
    const [input, updateInput] = useState({
        limit: 5,
        start: 0,
    });
    function updateInputValues(type, value) {
        updateInput({ ...input, [type]: value });
    }
    const [coins, updateCoins] = useState([]);
    const fetchCoins = useCallback(async () => {
        const { limit, start } = input;
        const data = await API.get(
            "cryptoapi",
            `/coins?limit=${limit}&start=${start}`
        );
        updateCoins(data.coins);
    }, [input]);
    useEffect(() => {
        fetchCoins();
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);
    return (
        <div className="App">
            <div>
                <input
                    onChange={(e) => updateInputValues("limit", e.target.value)}
                    placeholder="limit"
                    value={input.limit}
                />
                <input
                    onChange={(e) => updateInputValues("start", e.target.value)}
                    placeholder="start"
                    value={input.start}
                />
                <button onClick={fetchCoins}>Fetch Coins</button>
            </div>
            <div>
                {coins.map((coin, index) => (
                    <div key={index}>
                        <h2>
                            {coin.name} - {coin.symbol}
                        </h2>
                        <h5>${coin.price_usd}</h5>
                    </div>
                ))}
            </div>
        </div>
    );
}

export default App;
