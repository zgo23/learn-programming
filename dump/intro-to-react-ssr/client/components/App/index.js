import React from "react";
import style from "./style";

const App = () => (
    <>
        <div className={style.app}>Hello World</div>
        <button onClick={(e) => alert("Hello You!")}>Say Hello Back!</button>
    </>
);
export default App;
