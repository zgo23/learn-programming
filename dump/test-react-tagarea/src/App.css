html,
body,
#root {
    height: 100%;
}
.App {
    box-sizing: border-box;
    height: 100%;
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: center;
}
