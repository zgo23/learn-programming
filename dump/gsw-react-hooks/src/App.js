import "./App.css";
// import GetPreviouseValue from "./get-previous-value/example-2";
// import MeasureDomNode from "./measure-dom-node/example";
// import CounterUseReducer from "./counter-use-reducer/example";
import CounterUseReducer from "./counter-use-reducer/example-2";

function App() {
    return (
        <div className="App">
            <CounterUseReducer initialCount={10} />
        </div>
    );
}

export default App;
