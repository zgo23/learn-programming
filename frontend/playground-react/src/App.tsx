import React from "react";
// import StyledExample from "./styled-components-api/StyledExample";
// import TtlExample from "./styled-components-api/TtlExample";
// import MyName from "./react-hooks/basics/MyName";
// import MyName from "./react-hooks/basics/MyNameUseHook";
// import MyName from "./react-hooks/basics/MyNameUseSimulatedHook";
// import { BlogApp } from "./react-hooks/blog";
// import ErrorBoundaryExample from "./reactjs-org-docs/ErrorBoundaryExample";
// import HowContextsWork from "./react-hooks/how-contexts-work/HowContextsWorkApp";
// import ProblematicKeyDemo from "./reconciliation/element-key-prop/ProblematicKeyDemo";
// import CorrectKeyDemo from "./reconciliation/element-key-prop/CorrectKeyDemo";
// import A from "./reconciliation/abc/AbcDemo";
import EventDelegationDemo1 from "./event-delegation/demo1/index";
import "./App.css";

function App() {
    return (
        <div className="App">
            {/* <StyledExample />
            <TtlExample /> */}
            {/* <MyName /> */}
            {/* <BlogApp /> */}
            {/* <ErrorBoundaryExample /> */}
            {/* <HowContextsWork /> */}
            {/* <ProblematicKeyDemo /> */}
            {/* <CorrectKeyDemo /> */}
            {/* <A /> */}
            <EventDelegationDemo1 />
        </div>
    );
}

export default App;
