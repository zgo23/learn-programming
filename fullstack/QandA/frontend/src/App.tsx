import React from 'react';
import './App.css';
import { Header } from './Header';
import { HomePage } from './HomePage';

function App() {
  return (
    <div className="App">
      <Header />
      <HomePage />
    </div>
  );
}

export default App;
