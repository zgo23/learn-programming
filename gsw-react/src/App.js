import "./App.css";
// import SortControllerExample from "./better-functional-components/example";
import SortControllerExample from "./better-functional-components/example-3";

function App() {
    return (
        <div className="App">
            <SortControllerExample />
        </div>
    );
}

export default App;
