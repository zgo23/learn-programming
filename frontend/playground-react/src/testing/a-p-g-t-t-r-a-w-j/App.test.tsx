import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";

import App from "./App";
import Account from "./Account";
import { User } from "./types";

it("renders without crashing", () => {
  shallow(<App />);
});

it("renders Account header", () => {
  const wrapper = shallow(<App />);
  const welcome = <h1>Display Active Users Account Details</h1>;
  expect(wrapper.contains(welcome)).toEqual(true);
});

const user: User = {
  name: "Adeneye David",
  email: "david@gmail.com",
  username: "Dave",
};

describe("<Account />", () => {
  it("accepts user account props", () => {
    const wrapper = mount(<Account user={user} />);
    expect(wrapper.props().user).toEqual(user);
  });

  it("contains users account email", () => {
    const wrapper = mount(<Account user={user} />);
    const value = wrapper.find("p").text();
    expect(value).toEqual("david@gmail.com");
  });

  it("renders correctly with no error message", () => {
    const wrapper = mount(<App />);
    expect(wrapper.state("error")).toEqual(undefined);
  });
});

it("renders correctly", () => {
  const tree = shallow(<App />);
  expect(toJson(tree)).toMatchSnapshot();
});
