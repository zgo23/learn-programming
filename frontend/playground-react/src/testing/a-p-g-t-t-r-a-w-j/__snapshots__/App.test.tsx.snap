// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Fragment>
  <h1>
    Display Active Users Account Details
  </h1>
  <h3>
    Fetching Users....
  </h3>
</Fragment>
`;
